typing/typemod.ml

```diff
@@ -2260,59 +2267,83 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
 and type_application loc strengthen funct_body env smod =
   let rec extract_application funct_body env sargs smod =
     match smod.pmod_desc with
-    | Pmod_apply(f, sarg) ->
+    | Pmod_apply (f, sarg) ->
         let arg, shape = type_module true funct_body None env sarg in
-        let summary =
-          { loc=smod.pmod_loc;
-            attributes=smod.pmod_attributes;
-            f_loc = f.pmod_loc;
-            arg_is_syntactic_unit = sarg.pmod_desc = Pmod_structure [];
+        let summary = {
+          loc = smod.pmod_loc;
+          attributes = smod.pmod_attributes;
+          f_loc = f.pmod_loc;
+          arg = Some {
+            is_syntactic_unit = sarg.pmod_desc = Pmod_structure [];
             arg;
-            arg_path = path_of_module arg;
-            shape
+            path = path_of_module arg;
+            shape;
           }
-        in
+        } in
+        extract_application funct_body env (summary::sargs) f
+    | Pmod_apply_unit f ->
+        let summary = {
+          loc = smod.pmod_loc;
+          attributes = smod.pmod_attributes;
+          f_loc = f.pmod_loc;
+          arg = None
+        } in
         extract_application funct_body env (summary::sargs) f
     | _ -> smod, sargs
   in
   let sfunct, args = extract_application funct_body env [] smod in
   let funct, funct_shape =
-    let strengthen =
-      strengthen && List.for_all (fun {arg_path;_} -> arg_path <> None) args
+    let has_path { arg } = match arg with
+      | None | Some { path = None } -> false
+      | Some { path = Some _ } -> true
     in
+    let strengthen = strengthen && List.for_all has_path args in
     type_module strengthen funct_body None env sfunct
   in
   List.fold_left (type_one_application ~ctx:(loc, funct, args) funct_body env)
     (funct, funct_shape) args
 
 and type_one_application ~ctx:(apply_loc,md_f,args)
-    funct_body env (funct, funct_shape)  app_view =
+    funct_body env (funct, funct_shape) app_view =
   match Env.scrape_alias env funct.mod_type with
   | Mty_functor (Unit, mty_res) ->
-      if not app_view.arg_is_syntactic_unit then
-        raise (Error (app_view.f_loc, env, Apply_generative));
+      begin match app_view.arg with
+        | None -> ()
+        | Some arg ->
+          if arg.is_syntactic_unit then
+            Location.prerr_warning arg.arg.mod_loc
+              Warnings.Generative_application_expects_unit
```

Comment:
We could update the warning scope (with `Builtin_attributes.warning_scope`) here to ensure that
```ocaml
module M = F(struct end[@warning "-73"])
```
disables the warning. 

I was wondering if it might help `ppxlib` (but maybe ppxlib would have no choice other than disabling the globally warning whenever translating to or from a previous AST).

Reply:
If I am not missing anything, `ppxlib` might need to disable the warning on [`Ast_builder.pmod_apply`](https://ocaml.org/p/ppxlib/0.28.0/doc/Ppxlib/Ast_builder/Default/index.html#val-pmod_apply)-generated nodes, but won't need to disable it globally.

Ensuring that adding the attribute `[@warning "-73"]` on the functor argument disables the warning would be great if we decide to disable the warning on [`Ast_builder.pmod_apply`](https://ocaml.org/p/ppxlib/0.28.0/doc/Ppxlib/Ast_builder/Default/index.html#val-pmod_apply)-generated nodes.